File: experiment-7/brazil/runs/language=go_model=claude-opus-4-8-fast/rep2/loader.go
// Brazilian Soccer MCP Server
//
// File: loader.go
// Responsibility: Read the six bundled Kaggle CSV files and convert each row
// into the normalized `Match` / `Player` domain types. Each source file has a
// different column layout, so the loader maps columns by header name (tolerant
// of ordering and a UTF-8 BOM) and applies the dataset-specific competition
// label. All parsing funnels through the helpers in normalize.go so team names,
// dates and numbers come out canonicalized.
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
)

// matchFile describes how to load one of the match CSVs.
type matchFile struct {
	name        string // file name within the data dir
	competition string // competition label to stamp on every row (empty => per-row)
}

// LoadAll reads every dataset under dataDir into a freshly built Store. Missing
// files are skipped with a warning rather than being fatal, so the server still
// starts if a dataset is absent.
func LoadAll(dataDir string) (*Store, error) {
	s := NewStore()

	matchFiles := []matchFile{
		{"Brasileirao_Matches.csv", "Brasileirão Série A"},
		{"Brazilian_Cup_Matches.csv", "Copa do Brasil"},
		{"Libertadores_Matches.csv", "Copa Libertadores"},
		{"novo_campeonato_brasileiro.csv", "Brasileirão Série A"},
		{"BR-Football-Dataset.csv", ""}, // competition comes from the row
	}

	for _, mf := range matchFiles {
		path := filepath.Join(dataDir, mf.name)
		n, err := loadMatchFile(s, path, mf)
		if err != nil {
			fmt.Fprintf(os.Stderr, "warning: %s: %v\n", mf.name, err)
			continue
		}
		fmt.Fprintf(os.Stderr, "loaded %d matches from %s\n", n, mf.name)
	}

	playerPath := filepath.Join(dataDir, "fifa_data.csv")
	pn, err := loadPlayers(s, playerPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "warning: fifa_data.csv: %v\n", err)
	} else {
		fmt.Fprintf(os.Stderr, "loaded %d players from fifa_data.csv\n", pn)
	}

	s.Index()
	return s, nil
}

// newCSVReader opens path and returns a csv.Reader configured to tolerate the
// ragged rows present in some of the datasets.
func newCSVReader(path string) (*os.File, *csv.Reader, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, nil, err
	}
	r := csv.NewReader(f)
	r.FieldsPerRecord = -1 // allow variable column counts
	r.LazyQuotes = true
	r.ReuseRecord = true
	return f, r, nil
}

// headerIndex reads the first record and returns a map from normalized header
// name to column index. The leading UTF-8 BOM (present in fifa_data.csv) is
// stripped from the first column.
func headerIndex(r *csv.Reader) (map[string]int, error) {
	head, err := r.Read()
	if err != nil {
		return nil, err
	}
	idx := make(map[string]int, len(head))
	for i, h := range head {
		h = strings.TrimPrefix(h, "\ufeff")
		idx[strings.ToLower(strings.TrimSpace(h))] = i
	}
	return idx, nil
}

// get safely returns the trimmed value at the named column, or "" if the column
// is absent or the row is too short.
func get(row []string, idx map[string]int, col string) string {
	i, ok := idx[col]
	if !ok || i >= len(row) {
		return ""
	}
	return strings.TrimSpace(strings.Trim(row[i], `"`))
}

// loadMatchFile loads a single match CSV into the store and returns the number
// of rows ingested.
func loadMatchFile(s *Store, path string, mf matchFile) (int, error) {
	f, r, err := newCSVReader(path)
	if err != nil {
		return 0, err
	}
	defer f.Close()

	idx, err := headerIndex(r)
	if err != nil {
		return 0, err
	}

	count := 0
	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			continue // skip malformed line
		}
		m, ok := rowToMatch(row, idx, mf)
		if !ok {
			continue
		}
		s.Matches = append(s.Matches, m)
		count++
	}
	return count, nil
}

// rowToMatch converts one CSV row to a Match, returning false if the row lacks
// the minimum required fields (both team names).
func rowToMatch(row []string, idx map[string]int, mf matchFile) (Match, bool) {
	// Column names differ between the English Kaggle files and the Portuguese
	// novo_campeonato file; look up both spellings.
	home := firstNonEmpty(get(row, idx, "home_team"), get(row, idx, "home"), get(row, idx, "equipe_mandante"))
	away := firstNonEmpty(get(row, idx, "away_team"), get(row, idx, "away"), get(row, idx, "equipe_visitante"))
	if home == "" || away == "" {
		return Match{}, false
	}

	// States (read first so they can inform the canonical key for datasets that
	// keep the state in a separate column, e.g. novo_campeonato).
	homeState := firstNonEmpty(get(row, idx, "home_team_state"), get(row, idx, "mandante_uf"))
	awayState := firstNonEmpty(get(row, idx, "away_team_state"), get(row, idx, "visitante_uf"))

	m := Match{
		HomeTeam:  teamDisplay(home, homeState),
		AwayTeam:  teamDisplay(away, awayState),
		HomeKey:   teamFullKey(home, homeState),
		AwayKey:   teamFullKey(away, awayState),
		HomeBase:  teamBaseKey(home),
		AwayBase:  teamBaseKey(away),
		HomeState: homeState,
		AwayState: awayState,
		Source:    filepath.Base(mf.name),
	}

	// Scores.
	hg, okH := atoi(firstNonEmpty(get(row, idx, "home_goal"), get(row, idx, "gols_mandante")))
	ag, okA := atoi(firstNonEmpty(get(row, idx, "away_goal"), get(row, idx, "gols_visitante")))
	if okH && okA {
		m.HomeGoal, m.AwayGoal, m.HasScore = hg, ag, true
	}

	// Date: dedicated column, or a date+time pair (BR-Football-Dataset).
	dateRaw := firstNonEmpty(get(row, idx, "datetime"), get(row, idx, "data"), get(row, idx, "date"))
	if t, ok := parseDate(dateRaw); ok {
		m.Date, m.HasDate = t, true
	}

	// Season / year.
	if yr, ok := atoi(firstNonEmpty(get(row, idx, "season"), get(row, idx, "ano"))); ok {
		m.Season = yr
	} else if m.HasDate {
		m.Season = m.Date.Year()
	}

	// Round.
	m.Round = firstNonEmpty(get(row, idx, "round"), get(row, idx, "rodada"))

	// Stadium (historical Brasileirão only).
	m.Stadium = get(row, idx, "arena")

	// Stage (Libertadores only).
	m.Stage = get(row, idx, "stage")

	// Competition: fixed label, or per-row tournament (BR-Football-Dataset).
	if mf.competition != "" {
		m.Competition = mf.competition
	} else {
		m.Competition = competitionLabel(get(row, idx, "tournament"))
	}

	// Extended stats (BR-Football-Dataset).
	hs, okHS := atoi(get(row, idx, "home_shots"))
	as, okAS := atoi(get(row, idx, "away_shots"))
	hc, _ := atoi(get(row, idx, "home_corner"))
	ac, _ := atoi(get(row, idx, "away_corner"))
	if okHS || okAS {
		m.HomeShots, m.AwayShots = hs, as
		m.HomeCorners, m.AwayCorners = hc, ac
		m.HasStats = true
	}

	return m, true
}

// competitionLabel maps the BR-Football-Dataset "tournament" values to the
// canonical competition names used elsewhere.
func competitionLabel(tournament string) string {
	switch normKey(tournament) {
	case "serie a":
		return "Brasileirão Série A"
	case "serie b":
		return "Brasileirão Série B"
	case "serie c":
		return "Brasileirão Série C"
	case "copa do brasil":
		return "Copa do Brasil"
	case "":
		return "Unknown"
	default:
		return tournament
	}
}

// loadPlayers loads the FIFA player CSV into the store.
func loadPlayers(s *Store, path string) (int, error) {
	f, r, err := newCSVReader(path)
	if err != nil {
		return 0, err
	}
	defer f.Close()

	idx, err := headerIndex(r)
	if err != nil {
		return 0, err
	}

	count := 0
	for {
		row, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			continue
		}
		name := get(row, idx, "name")
		if name == "" {
			continue
		}
		p := Player{
			Name:        name,
			NameKey:     normKey(name),
			Nationality: get(row, idx, "nationality"),
			Club:        get(row, idx, "club"),
			ClubKey:     teamKey(get(row, idx, "club")),
			Position:    get(row, idx, "position"),
			Jersey:      get(row, idx, "jersey number"),
			Height:      get(row, idx, "height"),
			Weight:      get(row, idx, "weight"),
			PreferredFt: get(row, idx, "preferred foot"),
		}
		if id, ok := atoi(get(row, idx, "id")); ok {
			p.ID = id
		}
		if a, ok := atoi(get(row, idx, "age")); ok {
			p.Age = a
		}
		if o, ok := atoi(get(row, idx, "overall")); ok {
			p.Overall = o
		}
		if pt, ok := atoi(get(row, idx, "potential")); ok {
			p.Potential = pt
		}
		s.Players = append(s.Players, p)
		count++
	}
	return count, nil
}

// firstNonEmpty returns the first non-empty string among its arguments.
func firstNonEmpty(vals ...string) string {
	for _, v := range vals {
		if v != "" {
			return v
		}
	}
	return ""
}

File: experiments/adrianco/experiment-5/runs/language=typescript_model=claude-opus-4-7_tooling=beads/rep1/src/loader.ts
import { readFileSync } from 'node:fs';
import { resolve, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { parse } from 'csv-parse/sync';
import type { Match, Player, Competition } from './types.js';
import {
  normalizeTeamName,
  parseDate,
  parseNumber,
  parseOptionalNumber,
  parseSeason,
} from './normalize.js';

const __dirname = dirname(fileURLToPath(import.meta.url));

function defaultDataDir(): string {
  if (process.env.SOCCER_DATA_DIR) return process.env.SOCCER_DATA_DIR;
  return resolve(__dirname, '..', 'data', 'kaggle');
}

function readCsv(path: string): Record<string, string>[] {
  const raw = readFileSync(path, 'utf8');
  return parse(raw, {
    columns: true,
    skip_empty_lines: true,
    relax_quotes: true,
    relax_column_count: true,
    trim: true,
    bom: true,
  }) as Record<string, string>[];
}

function detectTournament(name: string): Competition {
  const n = name.toLowerCase();
  if (n.includes('libertadores')) return 'Copa Libertadores';
  if (n.includes('copa do brasil') || n.includes('brazilian cup')) return 'Copa do Brasil';
  if (n.includes('brasileir') || n.includes('serie a') || n.includes('série a')) return 'Brasileirão';
  return 'Other';
}

function buildBrasileirao(rows: Record<string, string>[]): Match[] {
  return rows.map((r): Match => {
    const date = parseDate(r.datetime);
    const season = parseSeason(r.season, date);
    return {
      competition: 'Brasileirão',
      source: 'Brasileirao_Matches.csv',
      date,
      season,
      round: r.round,
      homeTeamRaw: r.home_team,
      awayTeamRaw: r.away_team,
      homeTeam: normalizeTeamName(r.home_team),
      awayTeam: normalizeTeamName(r.away_team),
      homeTeamState: r.home_team_state || undefined,
      awayTeamState: r.away_team_state || undefined,
      homeGoals: parseNumber(r.home_goal),
      awayGoals: parseNumber(r.away_goal),
    };
  });
}

function buildCopaBrasil(rows: Record<string, string>[]): Match[] {
  return rows.map((r): Match => {
    const date = parseDate(r.datetime);
    const season = parseSeason(r.season, date);
    return {
      competition: 'Copa do Brasil',
      source: 'Brazilian_Cup_Matches.csv',
      date,
      season,
      round: r.round,
      homeTeamRaw: r.home_team,
      awayTeamRaw: r.away_team,
      homeTeam: normalizeTeamName(r.home_team),
      awayTeam: normalizeTeamName(r.away_team),
      homeGoals: parseNumber(r.home_goal),
      awayGoals: parseNumber(r.away_goal),
    };
  });
}

function buildLibertadores(rows: Record<string, string>[]): Match[] {
  return rows.map((r): Match => {
    const date = parseDate(r.datetime);
    const season = parseSeason(r.season, date);
    return {
      competition: 'Copa Libertadores',
      source: 'Libertadores_Matches.csv',
      date,
      season,
      stage: r.stage,
      homeTeamRaw: r.home_team,
      awayTeamRaw: r.away_team,
      homeTeam: normalizeTeamName(r.home_team),
      awayTeam: normalizeTeamName(r.away_team),
      homeGoals: parseNumber(r.home_goal),
      awayGoals: parseNumber(r.away_goal),
    };
  });
}

function buildExtended(rows: Record<string, string>[]): Match[] {
  return rows.map((r): Match => {
    const date = parseDate(r.date);
    const competition = detectTournament(r.tournament || '');
    const season = parseSeason(undefined, date);
    return {
      competition,
      source: 'BR-Football-Dataset.csv',
      date,
      season,
      homeTeamRaw: r.home,
      awayTeamRaw: r.away,
      homeTeam: normalizeTeamName(r.home),
      awayTeam: normalizeTeamName(r.away),
      homeGoals: parseNumber(r.home_goal),
      awayGoals: parseNumber(r.away_goal),
      stats: {
        homeCorners: parseOptionalNumber(r.home_corner),
        awayCorners: parseOptionalNumber(r.away_corner),
        totalCorners: parseOptionalNumber(r.total_corners),
        homeShots: parseOptionalNumber(r.home_shots),
        awayShots: parseOptionalNumber(r.away_shots),
        homeAttacks: parseOptionalNumber(r.home_attack),
        awayAttacks: parseOptionalNumber(r.away_attack),
        htHomeResult: r.ht_result || undefined,
        atAwayResult: r.at_result || undefined,
      },
    };
  });
}

function buildHistorical(rows: Record<string, string>[]): Match[] {
  return rows.map((r): Match => {
    const date = parseDate(r.Data);
    const season = parseSeason(r.Ano, date);
    return {
      competition: 'Brasileirão',
      source: 'novo_campeonato_brasileiro.csv',
      date,
      season,
      round: r.Rodada,
      homeTeamRaw: r.Equipe_mandante,
      awayTeamRaw: r.Equipe_visitante,
      homeTeam: normalizeTeamName(r.Equipe_mandante),
      awayTeam: normalizeTeamName(r.Equipe_visitante),
      homeTeamState: r.Mandante_UF || undefined,
      awayTeamState: r.Visitante_UF || undefined,
      homeGoals: parseNumber(r.Gols_mandante),
      awayGoals: parseNumber(r.Gols_visitante),
      arena: r.Arena || undefined,
    };
  });
}

function buildPlayers(rows: Record<string, string>[]): Player[] {
  return rows.map((r): Player => ({
    id: Number(r.ID),
    name: r.Name,
    age: parseOptionalNumber(r.Age),
    nationality: r.Nationality || undefined,
    overall: parseOptionalNumber(r.Overall),
    potential: parseOptionalNumber(r.Potential),
    club: r.Club || undefined,
    clubNormalized: r.Club ? normalizeTeamName(r.Club) : undefined,
    position: r.Position || undefined,
    jerseyNumber: parseOptionalNumber(r['Jersey Number']),
    height: r.Height || undefined,
    weight: r.Weight || undefined,
    preferredFoot: r['Preferred Foot'] || undefined,
    value: r.Value || undefined,
    wage: r.Wage || undefined,
    workRate: r['Work Rate'] || undefined,
    bodyType: r['Body Type'] || undefined,
  })).filter(p => p.name);
}

function shiftDate(date: string, days: number): string {
  if (!/^\d{4}-\d{2}-\d{2}/.test(date)) return date;
  const [y, m, d] = date.split('-').map(Number);
  const dt = new Date(Date.UTC(y, m - 1, d));
  dt.setUTCDate(dt.getUTCDate() + days);
  const yy = dt.getUTCFullYear();
  const mm = String(dt.getUTCMonth() + 1).padStart(2, '0');
  const dd = String(dt.getUTCDate()).padStart(2, '0');
  return `${yy}-${mm}-${dd}`;
}

function score(x: Match): number {
  return (x.stats ? 5 : 0) + (x.arena ? 1 : 0) + (x.round !== undefined && x.round !== '' ? 1 : 0) + (x.stage ? 1 : 0);
}

function dedupeMatches(matches: Match[]): Match[] {
  const map = new Map<string, Match>();
  // Try exact date first, then ±1 day fuzz against existing entries.
  for (const m of matches) {
    if (!m.date || !m.homeTeam || !m.awayTeam) continue;
    const baseKey = `${m.competition}|${m.homeTeam}|${m.awayTeam}|${m.homeGoals}-${m.awayGoals}`;
    const candidates = [
      `${baseKey}|${m.date}`,
      `${baseKey}|${shiftDate(m.date, -1)}`,
      `${baseKey}|${shiftDate(m.date, 1)}`,
    ];
    let existingKey: string | undefined;
    for (const key of candidates) {
      if (map.has(key)) { existingKey = key; break; }
    }
    if (!existingKey) {
      map.set(candidates[0], m);
      continue;
    }
    const existing = map.get(existingKey)!;
    if (score(m) > score(existing)) {
      // Keep existing key (don't move) but enrich with new fields.
      map.set(existingKey, {
        ...existing,
        round: existing.round ?? m.round,
        stage: existing.stage ?? m.stage,
        arena: existing.arena ?? m.arena,
        stats: existing.stats ?? m.stats,
        homeTeamState: existing.homeTeamState ?? m.homeTeamState,
        awayTeamState: existing.awayTeamState ?? m.awayTeamState,
      });
    } else {
      // Existing wins — but fill in missing fields from candidate
      map.set(existingKey, {
        ...existing,
        round: existing.round ?? m.round,
        stage: existing.stage ?? m.stage,
        arena: existing.arena ?? m.arena,
        stats: existing.stats ?? m.stats,
        homeTeamState: existing.homeTeamState ?? m.homeTeamState,
        awayTeamState: existing.awayTeamState ?? m.awayTeamState,
      });
    }
  }
  return Array.from(map.values());
}

export interface SoccerData {
  matches: Match[];
  players: Player[];
  dataDir: string;
}

export function loadData(dataDir?: string): SoccerData {
  const dir = dataDir || defaultDataDir();
  const matches: Match[] = [];
  matches.push(...buildBrasileirao(readCsv(resolve(dir, 'Brasileirao_Matches.csv'))));
  matches.push(...buildCopaBrasil(readCsv(resolve(dir, 'Brazilian_Cup_Matches.csv'))));
  matches.push(...buildLibertadores(readCsv(resolve(dir, 'Libertadores_Matches.csv'))));
  matches.push(...buildExtended(readCsv(resolve(dir, 'BR-Football-Dataset.csv'))));
  matches.push(...buildHistorical(readCsv(resolve(dir, 'novo_campeonato_brasileiro.csv'))));
  const players = buildPlayers(readCsv(resolve(dir, 'fifa_data.csv')));
  return { matches: dedupeMatches(matches), players, dataDir: dir };
}

let cached: SoccerData | undefined;
export function getData(dataDir?: string): SoccerData {
  if (!cached || (dataDir && dataDir !== cached.dataDir)) {
    cached = loadData(dataDir);
  }
  return cached;
}
